plasmapy/atomic/tests/test_ionization_state.py

```diff
@@ -0,0 +1,297 @@
+import pytest
+import numpy as np
+import astropy.units as u
+from ..ionization_states import IonizationState
+from ...utils import AtomicError, RunTestError, InvalidIsotopeError, run_test
+from ...atomic import (
+    atomic_number,
+    mass_number,
+    atomic_symbol,
+    isotope_symbol,
+    Particle,
+)
+import collections
+import warnings
+
+
+test_cases = {
+    'Li': {
+        'particle': 'Li',
+        'ionic_fractions': np.array([0.4, 0.3, 0.2, 0.1]),
+        'tol': 1e-15,
+    },
+
+    'Li ground state': {
+        'particle': 'Li',
+        'ionic_fractions': np.array([1, 0, 0, 0], dtype=np.int64),
+        'tol': 1e-15,
+    },
+
+    'H': {
+        'particle': 'H',
+        'ionic_fractions': [0.6, 0.4],
+        'tol': 1e-8,
+    },
+
+    'H acceptable error': {
+        'particle': 'H',
+        'ionic_fractions': [0.6, 0.400_000_001],
+        'tol': 1e-8,
+    },
+
+    'D': {
+        'particle': 'deuterium',
+        'ionic_fractions': [0.7, 0.3],
+        'tol': 1e-15,
+        'n_e': 3e14 * u.m ** -3,
+    },
+
+    'He': {
+        'particle': 'He',
+        'ionic_fractions': [0.5, 0.3, 0.2],
+        'n_elem': 1e20 * u.m ** -3,
+    },
+
+    'number densities': {
+        'particle': 'T',
+        'ionic_fractions': np.array([1e4, 1]) * u.cm ** -3,
+    }
+
+}
+
+test_names = test_cases.keys()
+
+
+class Test_IonizationState:
+    """Test instances of IonizationState."""
+
+    @classmethod
+    def setup_class(cls):
+        cls.instances = {}
+
+    @pytest.mark.parametrize('test_name', test_names)
+    def test_instantiation(self, test_name):
+        try:
+            self.instances[test_name] = IonizationState(**test_cases[test_name])
+        except Exception as exc:
+            raise RunTestError(
+                f"Unable to create IonizationState instance for "
+                f"test case {test_name}.")
+
+    @pytest.mark.parametrize(
+        'test_name',
+        [name for name in test_names if 'ionic_fractions' in test_cases[name].keys()],
+    )
+    def test_ionic_fractions(self, test_name):
+        instance = self.instances[test_name]
+
+        inputted_fractions = test_cases[test_name]['ionic_fractions']
+        if isinstance(inputted_fractions, u.Quantity):
+            inputted_fractions = inputted_fractions.to(u.m ** -3)
+            inputted_fractions = (inputted_fractions / inputted_fractions.sum()).value
+
+        try:
+            assert np.allclose(instance.ionic_fractions, inputted_fractions)
+        except:
```

Comment:
I am guessing, we're going to catch (only) `AssertionError` here. If so, it would be a good idea if we specifically mention that (`except AssertionError: ...`). We could also display values for `instance.ionic_fractions` and `inputted_fractions` when re-throwing the error. :blush: